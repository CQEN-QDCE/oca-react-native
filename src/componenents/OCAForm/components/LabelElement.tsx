import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import type { LabelElementProps } from '../../types';

export const LabelElement = ({ attribute, style }: LabelElementProps) => {
  return (
    <View>
      <Text style={[styles.label, style]} numberOfLines={1}>
        {attribute.name}
      </Text>
    </View>
  );
};

const styles = StyleSheet.create({
  label: {
    color: '#333',
    fontWeight: 'bold',
    fontSize: 14,
  },
});
