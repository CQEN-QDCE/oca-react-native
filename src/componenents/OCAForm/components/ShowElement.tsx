import React from 'react';
import { TouchableOpacity, View, Text } from 'react-native';
import type { ShowElementProps } from '../../types';
import { attributeTypeStyle } from '../../utils/attributeTypeStyle';

export const ShowElement = ({
  hideShowOptions,
  onToggleViewPressed,
  shown = false,
  style,
}: ShowElementProps) => {
  return (
    <View>
      {hideShowOptions.visibility && (
        <TouchableOpacity activeOpacity={1} onPress={onToggleViewPressed}>
          <View>
            {shown ? (
              hideShowOptions?.labelHide ? (
                hideShowOptions.labelHide
              ) : (
                <Text style={[attributeTypeStyle.text, style]}>Hide</Text>
              )
            ) : hideShowOptions?.labelShow ? (
              hideShowOptions.labelShow
            ) : (
              <Text style={[attributeTypeStyle.text, style]}>Show</Text>
            )}
          </View>
        </TouchableOpacity>
      )}
    </View>
  );
};
