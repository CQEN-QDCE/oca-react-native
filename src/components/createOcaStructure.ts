import type { OCA } from 'oca.js';
import { OcaJs } from '../packages/oca.js-form-core/OcaJs';

export const createOcaStructure = async (oca?: OCA) => {
  const ocaJs = new OcaJs({});
  if (oca) {
    return await ocaJs.createStructure(oca);
  }
  return undefined;
};
