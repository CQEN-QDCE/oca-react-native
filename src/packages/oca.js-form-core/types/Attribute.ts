import type { AttributeTranslation } from './AttributeTranslation';
import type { Translations } from './Translations';

export type Attribute = {
  characterEncoding?: string;
  entryCodes?: string[];
  entryCodesMapping?: string[];
  format?: string;
  metric_system?: string;
  unit?: string;
  sai?: string;
  condition?: string;
  dependencies?: string[];
  mapping?: string;
  cardinality?: string;
  conformance?: 'O' | 'M';
  translations: Translations<AttributeTranslation>;
};
