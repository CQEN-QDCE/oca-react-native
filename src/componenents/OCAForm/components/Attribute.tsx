import React from 'react';
import type { AttributeProps } from '../types';
import { StyleSheet, View } from 'react-native';
import { LabelElement } from './LabelElement';
import { ShowElement } from './ShowElement';
import { TextAttribute } from './AttributesTypes/TextAttribute';
import { DateAttribute } from './AttributesTypes/DateAttribute';
import { SelectAttribute } from './AttributesTypes/SelectAttribute';
import { BinaryAttribute } from './AttributesTypes/BinaryAttribute';

export const Attribute = ({
  attribute,
  shown,
  hideShowOptions,
  onToggleViewPressed,
  stylingOptions,
  attributeLabel = null,
  attributeValue = null,
}: AttributeProps) => {
  const attributeTypeSelector = () => {
    if (attribute.type === 'Date') {
      console.log('DATE VALUE: ' + attribute.name + ' - ' + attribute.value);
      return (
        <DateAttribute
          attribute={attribute}
          shown={shown}
          styles={stylingOptions?.textStyle}
        />
      );
    } else if (attribute.type === 'Select') {
      return (
        <SelectAttribute
          attribute={attribute}
          shown={shown}
          styles={stylingOptions?.textStyle}
        />
      );
    } else if (attribute.type === 'Binary') {
      return (
        <BinaryAttribute
          attribute={attribute}
          shown={shown}
          binaryStyle={stylingOptions?.binaryStyle}
          textStyle={stylingOptions?.textStyle}
        />
      );
    }
    return (
      <TextAttribute
        attribute={attribute}
        shown={shown}
        styles={stylingOptions?.textStyle}
      />
    );
  };

  return (
    <View style={[styles.container, stylingOptions?.attributeContainerStyle]}>
      {attributeLabel ? (
        attributeLabel(attribute)
      ) : (
        <LabelElement
          attribute={attribute}
          style={stylingOptions?.labelTextStyle}
        />
      )}

      {attributeValue ? (
        attributeValue(attribute)
      ) : (
        <View style={styles.valueContainer}>
          {attributeTypeSelector()}

          <ShowElement
            hideShowOptions={hideShowOptions}
            shown={shown}
            onToggleViewPressed={onToggleViewPressed}
          />
        </View>
      )}
      <View style={[styles.separator, stylingOptions?.separatorStyle]} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    paddingHorizontal: 25,
    paddingTop: 16,
    backgroundColor: 'transparent',
  },
  valueContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingTop: 10,
    maxWidth: '100%',
  },
  separator: {
    borderBottomColor: '#c9c9c9',
    borderBottomWidth: 2,
    paddingTop: 12,
  },
});
