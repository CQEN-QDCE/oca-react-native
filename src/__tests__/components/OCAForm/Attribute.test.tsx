import React from 'react';
import renderer from 'react-test-renderer';
import { Attribute } from '../../../components/OCAForm/components/Attribute';
import { Text } from 'react-native';
const textAttribute = { name: 'Birth date', value: '2000-02-21', type: 'Text' };
const binaryAttribute = {
  name: 'Picture',
  value:
    'iVBORw0KGgoAAAANSUhEUgAAAV4AAAFeAQAAAADlUEq3AAAAAW9yTlQBz6J3mgAAACZJREFUaN7twTEBAAAAwqD1T+1pCaAAAAAAAAAAAAAAAAAAAAC4AT2GAAGWvJzxAAAAAElFTkSuQmCC',
  type: 'Binary',
};

const onTogglePress = () => {};

describe('Oca Attribute', () => {
  it('Given a text attribute should render properly.', () => {
    const tree = renderer
      .create(
        <Attribute
          attribute={textAttribute}
          onToggleViewPressed={onTogglePress}
          shown={true}
          hideShowOptions={{ visibility: true }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Given a binary attribute should render properly', () => {
    const tree = renderer
      .create(
        <Attribute
          attribute={binaryAttribute}
          onToggleViewPressed={onTogglePress}
          shown={true}
          hideShowOptions={{ visibility: true }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Given a attributeLabel should render the attribute label properly', () => {
    const tree = renderer
      .create(
        <Attribute
          attribute={textAttribute}
          onToggleViewPressed={onTogglePress}
          shown={true}
          hideShowOptions={{ visibility: false }}
          attributeLabel={(attribute) => {
            return <Text>{attribute.name}</Text>;
          }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Given a attributeValue should render the attribute value properly', () => {
    const tree = renderer
      .create(
        <Attribute
          attribute={textAttribute}
          onToggleViewPressed={onTogglePress}
          shown={true}
          hideShowOptions={{ visibility: false }}
          attributeValue={(attribute) => {
            return <Text>{attribute.value}</Text>;
          }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
