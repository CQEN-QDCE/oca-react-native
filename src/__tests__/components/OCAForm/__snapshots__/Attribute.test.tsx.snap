// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Oca Attribute Given a attributeLabel should render the attribute label properly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "paddingHorizontal": 25,
        "paddingTop": 16,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <Text>
    Birth date
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "maxWidth": "100%",
        "paddingTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "color": "#333",
          "paddingVertical": 4,
        }
      }
    >
      <Text
        numberOfLines={200}
        style={
          Array [
            Object {
              "color": "#333",
              "fontSize": 18,
              "fontWeight": "normal",
            },
            undefined,
          ]
        }
      >
        2000-02-21
      </Text>
    </View>
    <View />
  </View>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "#c9c9c9",
          "borderBottomWidth": 2,
          "paddingTop": 12,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Oca Attribute Given a attributeValue should render the attribute value properly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "paddingHorizontal": 25,
        "paddingTop": 16,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#333",
            "fontSize": 14,
            "fontWeight": "bold",
          },
          undefined,
        ]
      }
    >
      Birth date
    </Text>
  </View>
  <Text>
    2000-02-21
  </Text>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "#c9c9c9",
          "borderBottomWidth": 2,
          "paddingTop": 12,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Oca Attribute Given a binary attribute should render properly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "paddingHorizontal": 25,
        "paddingTop": 16,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#333",
            "fontSize": 14,
            "fontWeight": "bold",
          },
          undefined,
        ]
      }
    >
      Picture
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "maxWidth": "100%",
        "paddingTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "color": "#333",
          "paddingVertical": 4,
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "iVBORw0KGgoAAAANSUhEUgAAAV4AAAFeAQAAAADlUEq3AAAAAW9yTlQBz6J3mgAAACZJREFUaN7twTEBAAAAwqD1T+1pCaAAAAAAAAAAAAAAAAAAAAC4AT2GAAGWvJzxAAAAAElFTkSuQmCC",
          }
        }
        style={
          Object {
            "aspectRatio": 1,
            "borderRadius": 10,
            "height": 150,
            "resizeMode": "contain",
          }
        }
      />
    </View>
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#333",
                  "fontSize": 18,
                  "fontWeight": "normal",
                },
                undefined,
              ]
            }
          >
            Hide
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "#c9c9c9",
          "borderBottomWidth": 2,
          "paddingTop": 12,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Oca Attribute Given a text attribute should render properly. 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "paddingHorizontal": 25,
        "paddingTop": 16,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#333",
            "fontSize": 14,
            "fontWeight": "bold",
          },
          undefined,
        ]
      }
    >
      Birth date
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "maxWidth": "100%",
        "paddingTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "color": "#333",
          "paddingVertical": 4,
        }
      }
    >
      <Text
        numberOfLines={200}
        style={
          Array [
            Object {
              "color": "#333",
              "fontSize": 18,
              "fontWeight": "normal",
            },
            undefined,
          ]
        }
      >
        2000-02-21
      </Text>
    </View>
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#333",
                  "fontSize": 18,
                  "fontWeight": "normal",
                },
                undefined,
              ]
            }
          >
            Hide
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "#c9c9c9",
          "borderBottomWidth": 2,
          "paddingTop": 12,
        },
        undefined,
      ]
    }
  />
</View>
`;
